### PR TITLE
Add MGLCircle (with radius expressed in physical units) 

### LINKED ISSUE #2167: Add MGLCircle (with radius expressed in physical units)
Now that v8 has landed in https://github.com/mapbox/mapbox-gl-native/commit/0451aca37e866c3380072b114e2d49353888e4b5 we have a `Circle` layer type.

### PR DESCRIPTION
**Pull request by 1ec5**
A new MGLCircle class, analogous to MapKit’s [MKCircle](https://developer.apple.com/documentation/mapkit/mkcircle) or the Google Map SDK’s [GMSCircle](https://developers.google.com/maps/documentation/ios-sdk/shapes#circles), allows the developer to add [spherical caps](https://en.wikipedia.org/wiki/Spherical_cap) to the map view as overlays or to a shape source as polygons.

Unlike MGLCircleStyleLayer or the Android annotation plugin’s [`Circle`](https://github.com/mapbox/mapbox-plugins-android/blob/master/plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/Circle.java), spherical caps are measured in meters, don’t necessarily _appear_ circular in the Web Mercator projection, and scale up to the whole world regardless of the current zoom level. Unlike [turf-circle](https://github.com/Turfjs/turf/tree/master/packages/turf-circle/) or #14429, this implementation is specific to the Web Mercator projection, the number of sides is chosen automatically, and it’s possible for a spherical cap to envelop the north or south pole. Circles still aren’t quite a first-class geometry type: GeoJSON doesn’t have a way to express curves, so under the hood, a circle is approximated as a many-sided regular polygon.

boston-light<br>_Concentric circle annotations with a one-line pulsing animation._

boston-harbor-islands<br>_These circles remain visible even though their centers are out of view. Hit testing works throughout the circles._

daylight<br>_Day and night (and twilight), implemented as a series of concentric circles with radii equal to about a quarter of the Earth’s circumference._

MGLCircle is optimized for accuracy, with some consideration for performance. The number of sides in an MGLCircle’s underlying polygon depends on the circle’s radius and center latitude. This class relies on mbgl to perform Douglas–Peucker simplification on the shape at the tolerance appropriate to the map’s current zoom level. The class tries to choose enough sides that the internal angles are imperceptibly slight, but not so many that the extra detail is lost when mbgl simplifies the shape at the highest possible zoom level. The simplification heuristic determines the number of sides until the radius reaches about 10&nbsp;kilometers; beyond that, angle heuristic caps the number of sides at 21,600. Despite these optimizations, there is a tradeoff in terms of performance, especially when the application doesn’t need much detail at all.

MGLCircle primarily targets the annotation API because only the annotation API is capable of persisting the circle as a circle (#6178, #7376), which makes it possible for the developer to modify the center coordinate and radius on the fly (#6177) without explicitly rebuilding the entire polygon from scratch. Most of the functionality introduced in this PR will remain relevant even after the annotation API is reimplemented atop the runtime styling API (#6181, #12863), though some operations will likely become less performant or more unwieldy internally. For now, I’m labeling this PR as <kbd>DO NOT MERGE</kbd> in case @mapbox/maps-ios has concerns about the new annotation API’s ability to maintain compatibility with this new annotation type, but it is already functional.

The Debug ‣ Add Polyline and Polygon Annotations menu item in macosapp has been replaced by a Simulate Nightfall menu item that turns the map into a daylight chart (including the different degrees of twilight) and adds a pulsing light that radiates out from [Boston Light](https://en.wikipedia.org/wiki/Boston_Light), one of the oldest lighthouses in the United States.

Geodesic polylines (#1726) could be implemented using a similar approach as this PR.

Fixes #2167.

### CODE REVIEW

**Comment by 1ec5** [comment]
> allows the developer to add spherical caps … to a shape source as polygons

To clarify, it is possible to use MGLCircle with the runtime styling API, by putting one in an MGLShapeSource. However, it isn’t possible to preserve any metadata associated with the circle when storing it that way. Until #7454 is fixed, we’d either need to define an MGLCircleFeature class or an MGLPolygonFeature initializer that accepts an MGLCircle.

> MGLCircle primarily targets the annotation API because only the annotation API is capable of persisting the circle as a circle (#6178, #7376), which makes it possible for the developer to modify the center coordinate and radius on the fly (#6177) without explicitly rebuilding the entire polygon from scratch.

To expand on this point: in theory, MGLCircleFeature or MGLPolygonFeature could stash MGLCircle’s `coordinate` and `radius` properties in attributes (namespaced to avoid collisions with user-defined attributes), and any conversion from GeoJSON to MGLShape objects could use these attributes to reconstruct an MGLCircle instead of an MGLPolygon. Unfortunately, even shape sources get tiled up (#6178), potentially splicing the circle into a bunch of irregular polygons that vary by zoom level. It would be difficult, though possible, to reconstruct a single circle based on the attributes saved from the original MGLCircle, then manually cut it back up into tiles.

**Comment by julianrex** [comment]
@asheemmamoowala a while back you had suggested using computed shape sources for this - can you comment on why that might be preferable?

**Comment by 1ec5** [comment]
I did consider a computed shape source as a solution to this problem but decided against it. In theory, for large circles, such as the nighttime spherical cap, a computed shape source would only have to generate visible portions of the circle, which could cut down on memory and CPU usage, and it could generate only as many points as needed for the current zoom level. But generating only the visible portion of the circle would require nontrivial tile cover logic.

It turns out that mbgl is already pretty efficient at both simplifying and tiling shapes, so I’d contend that it’s better to generate the entire circle upfront and let mbgl tile and simplify it dynamically as the zoom level changes. This keeps the SDK implementation lightweight, leveraging existing functionality in mbgl.

In effect, we’re optimizing for the user panning and zooming rather than statically viewing only a small portion of the circle. If this isn’t the right tradeoff for a particular developer’s use case, they’re free to implement a computed shape source on their own.

**Comment by asheemmamoowala** [comment]
> generating only the visible portion of the circle would require nontrivial tile cover logic that I’m uncomfortable implementing at the SDK level.

This is a good reason not to use a custom source. If we wanted to support this on all platforms - it might still be worth investigating - but the cost of tiling will still be non-trivial.

File: platform/macos/app/Base.lproj/MainMenu.xib [hunk 1]
```diff
@@ -542,9 +542,9 @@
                                     <action selector="dropManyPins:" target="-1" id="Rtv-8N-3Z8"/>
                                 </connections>
                             </menuItem>
-                            <menuItem title="Add Polygon and Polyline" keyEquivalent="l" id="DVr-vT-lpe">
+                            <menuItem title="Simulate Nightfall" keyEquivalent="l" id="DVr-vT-lpe">
```
Comment (asheemmamoowala) [inline platform/macos/app/Base.lproj/MainMenu.xib]: nit: Is this related to this PR?

Comment (asheemmamoowala) [inline platform/macos/app/Base.lproj/MainMenu.xib]: nit: Can you add a subtitle/tooltip that this item uses Spherical caps or `MGLCircle`

File: platform/darwin/src/MGLCircle.h [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ circle is defined by a center coordinate, specified as a
+ `CLLocationCoordinate2D` instance, and a physical radius measured in meters.
+ The circle is approximated as a polygon with a large number of vertices and
+ edges. Due to the map’s Spherical Mercator projection, a large enough circle
+ appears as an elliptical or even sinusoidal shape. You could use a circle to
+ visualize for instance an impact zone, the `CLLocation.horizontalAccuracy` of a
+ GPS location update, the regulated airspace around an airport, or the
+ addressable consumer market around a city.
+
+ You can add a circle overlay directly to a map view using the
```
Comment (asheemmamoowala) [inline platform/darwin/src/MGLCircle.h]: Do you think it is necessary to call out that this type does not support all the same paint/layout options as a shape added to an `MGLFillStyleLayer`?  I'm thinking specifically of fill patterns, but there are probably others as well.

**Comment by jlubeck** [comment]
What happened to this PR? Was so looking forward to it!

**Comment by 1ec5** [comment]
This PR will need to be moved over to the [mapbox-gl-native-ios](https://github.com/mapbox/mapbox-gl-native-ios/) repository, now that the iOS and macOS map SDKs are maintained there. I’d be happy to do so, once there’s a decision on whether this feature will be accepted: https://github.com/mapbox/mapbox-gl-native/issues/2167#issuecomment-519221772.

If `-[MGLMapView addAnnotations:]` and platform-specific annotation functionality is removed at any time before this feature lands, then MGLCircle itself will continue to exist, but it would need to be used through MGLShapeSource, which converts the shape into a many-sided polygon with a fixed radius.

**Comment by bsudekum** [comment]
The people need MGLCircle.
